contracts/RewardManager.sol

```diff
@@ -180,36 +222,90 @@ contract RewardManager is Ownable, Versionable, Safe {
     address payable rewardSafe,
     address rewardProgramID
   ) public view returns (bool) {
-    address rewardSafeOwner = getRewardSafeOwner(rewardSafe);
-    return rewardSafe == rewardSafes[rewardProgramID][rewardSafeOwner];
+    return rewardSafes[rewardProgramID].contains(rewardSafe);
   }
 
-  // External View Functions
-  function hasRule(address rewardProgramID, string calldata ruleDID)
-    external
+  function encodeTransactionData(bytes memory signature)
+    public
     view
-    returns (bool)
+    returns (address, bytes memory)
   {
-    if (_equalRule(rule[rewardProgramID][ruleDID], Rule("", ""))) {
-      return false;
-    } else {
-      return true;
-    }
+    (
+      address to,
+      uint256 value,
+      bytes memory payload,
+      uint8 operation,
+      uint256 safeTxGas,
+      uint256 baseGas,
+      uint256 gasPrice,
+      address gasToken,
+      address refundReceiver,
+      uint256 nonce
+    ) =
+      abi.decode(
+        signature,
+        (
+          address,
+          uint256,
+          bytes,
+          uint8,
+          uint256,
+          uint256,
+          uint256,
+          address,
+          address,
+          uint256
+        )
+      );
+    return (
+      to,
+      GnosisSafe(msg.sender).encodeTransactionData(
+        to,
+        value,
+        payload,
+        Enum.Operation.Call,
+        safeTxGas,
+        baseGas,
+        gasPrice,
+        gasToken,
+        refundReceiver,
+        nonce
+      )
+    );
+  }
+
+  function isValidSignature(bytes memory data, bytes memory signature)
+    public
+    view
+    returns (bytes4)
+  {
+    (address to, bytes memory encodedTransactionData) =
+      encodeTransactionData(signature);
+    address rewardSafeOwner = getRewardSafeOwner(msg.sender);
+
+    bytes memory contractSignature =
+      _contractSignature(msg.sender, rewardSafeOwner);
+    return
+      ((_equalBytes(data, encodedTransactionData) &&
+        (to == msg.sender && signatures[keccak256(contractSignature)])) ||
+        (to == address(this)) ||
+        (eip1271Contracts.contains(to)))
```

Comment:
and even more nuance (which is what I think you were getting to in your comments for the PR), is that for the `transferRewardSafes`, because that is executed by merit of a safe txn (because we are dealing with nested safe txns), to even invoke that we need to allow the reward manager permit that to be called. But our safety valve is that we have specifically permitted that txn to be performed because we have included that safe txn's arguments as the signature itself (which includes nonce to prevent replays)

Reply:
all in all, i think this is fine, I'm trying to think of a way to hack this, but I think we have all our bases covered. I think a nice detailed explanation, though, is necessary for posterity.

Reply:
The reward safe has two owners: eoa and reward manager contract <- this is for certain.

Reply:
Anyway, think this may need a deeper dive. I'll merge and then resolve. Think what needs to be clear is what are we using `isValidSignature` for. Maybe more context will be found with a specific use-case for reward safe like in a market and commerce context. 

> Ownership related safe functions, though, we absolutely want locked down (as well as any other safe mutating function--i don't know others off the top of my head). because we require the additional 2nd signature (that we obtained via the transferRewardSafe function), ownership related changes like add owner, remove owner, change threshold, will be blocked.

noted